packages/dds/tree/docs/wip/Schema.ts

```diff
@@ -0,0 +1,621 @@
+ *
+ * This intentionally does not support any concept of versioning or name-spacing:
+ * users of it can include versions and namespaces in if they want, or features for them could be added later.
+ *
+ * Can be used either:
+ * 1. Record just this in the persisted data.
+ * When loading the data, see if the loader has a schema with a matching name, and if so, use that.
+ * Optionally validate loaded data against schema.
+ * 2. Persist the whole schema. Use the identifier to associate it with schema when loading to check that the schema match.
+ */
+export type SchemaIdentifier = string & {
+    readonly SchemaIdentifier: "3965006b-66d0-41d6-bb04-bd6873b528b4";
+};
+
+/**
+ * SchemaIdentifier for a Tree.
+ * Also known as "Definition"
+ */
+export type TreeSchemaIdentifier = SchemaIdentifier & {
+    readonly TreeSchemaIdentifier: "ffc4b4b6-a4d8-4479-9636-fe6c6a1a4a7f";
+};
+
+/**
+ * Key / Name / Label for a field which is scoped to a specific TreeSchema.
+ */
+export type LocalFieldKey = string & {
+    readonly TreeSchemaIdentifier: "1108736b-ebb5-4924-9a25-d4dbabd83fc5";
+};
+
+/**
+ * SchemaIdentifier for a Field "global field",
+ * meaning a field which has the same meaning for all usages withing the document (not scoped to a specific TreeSchema like LocalFieldKey).
+ *
+ * Note that the implementations should ensure that GlobalFieldKeys can never collide with LocalFieldKeys.
+ * This can either be done in several ways
+ * (keeping the two classes of fields separate, namespaceing/escaping, compressing one into numbers and leaving the other strings, etc.)
+ */
+export type GlobalFieldKey = SchemaIdentifier & {
+    readonly TreeSchemaIdentifier: "7cc802f9-3360-4176-bbbf-cab23320d80b";
+};
+
+/**
+ * Schema for how many children are allowed in a field.
+ */
+export enum Multiplicity {
+    /**
+     * Exactly one item.
+     */
+    Value,
+    /**
+     * 0 or 1 items.
+     */
+    Optional,
+    /**
+     * 0 or more items.
+     */
+    Sequence,
+    /**
+     * Exactly 0 items.
+     */
+    Forbidden,
+}
+
+/**
+ * Example for how we might want to handle values.
+ *
+ * This might be significantly different if we want to focus more on binary formats (need to work out how fluid GC would work with that).
+ * For now, this is a simple easy to support setup.
+ *
+ * Note that use of non-Nothing values might be restricted in the actual user facing schema languages:
+ * we could instead choose to get by with the only types supporting values being effectively builtin,
+ * though this limitation could prevent users for updating/extending the primitive schema to allow the annotations they might want.
+ *
+ * An interesting alternative to this simple value Enum would be to use something more expressive here, like JsonSchema:
+ * since this is modeling immutable data, we really just need a way to figure out which if these value schema allow super sets of each-other.
+ */
+export enum ValueSchema {
```

Comment:
Nit: I think it's more natural to refer to these as ValueType, FieldType, etc.

Both GraphQL and JSON schema use the term "type" for these concepts.